text/0034-ternary-hash.md

```diff
@@ -0,0 +1,472 @@
+  let mut curlp = CurlP81::new();
+
+  // Assume there are some transaction trits, all zeroes for the sake of this example.
+  let transaction = TritBuf::<T1B1Buf>::zeros(6561);
+  let mut hash = TritBuf::<T1B1Buf>::zeros(243);
+
+  // Absorb the transaction.
+  curlp.absorb(&transaction);
+  // Squeeze out a hash.
+  curlp.squeeze_into(&mut hash);
+  ```
+
+- `kerl`
+  ```rust
+  // Create a Kerl instance.
+  let mut kerl = Kerl::new();
+
+  // `Kerl::digest` is a function that combines `Kerl::absorb` and `Kerl::squeeze`.
+  // `Kerl::digest_into` combines `Kerl::absorb` with `Kerl::squeeze_into`.
+  let hash = kerl.digest(&transaction);
+  ```
+
+The main proposal of this RFC are the `Sponge` trait and the `CurlP` and `Kerl` types that are implementing it.
+This RFC relies on the presence of the types `TritBuf` and `Trits`, as defined by
+[RFC36](https://github.com/iotaledger/bee-rfcs/blob/master/text/0036-ternary.md), which are assumed to be owning and
+borrowing collections of binary-encoded ternary in the `T1B1` encoding (one trit per byte).
+
+```rust
+/// The common interface of cryptographic hash functions that follow the sponge construction and that act on ternary.
+trait Sponge {
+    /// An error indicating that a failure has occured during a sponge operation.
+    type Error;
+
+    /// Absorb `input` into the sponge.
+    fn absorb(&mut self, input: &Trits) -> Result<(), Self::Error>;
+
+    /// Reset the inner state of the sponge.
+    fn reset(&mut self);
+
+    /// Squeeze the sponge into a buffer
+    fn squeeze_into(&mut self, buf: &mut Trits) -> Result<(), Self::Error>;
+
+    /// Convenience function using `Sponge::squeeze_into` to return an owned version of the hash.
+    fn squeeze(&mut self) -> Result<TritBuf, Self::Error> {
+        let mut output = TritBuf::zeros(HASH_LENGTH);
+        self.squeeze_into(&mut output)?;
+        Ok(output)
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge into a buffer, and reset the sponge in one go.
+    fn digest_into(&mut self, input: &Trits, buf: &mut Trits) -> Result<(), Self::Error> {
+        self.absorb(input)?;
+        self.squeeze_into(buf)?;
+        self.reset();
+        Ok(())
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge, and reset the sponge in one go.
+    /// Returns an owned version of the hash.
+    fn digest(&mut self, input: &Trits) -> Result<TritBuf, Self::Error> {
+        self.absorb(input)?;
+        let output = self.squeeze()?;
+        self.reset();
+        Ok(output)
+    }
+}
+```
+
+Following the sponge metaphor, an input provided by the user is `absorb`ed, and an output will be `squeeze`d from the
+data structure. `digest` is a convenience method calling `absorb` and `squeeze` in one go. The `*_into` versions of
+these methods are for passing a buffer into which the calculated hashes are written. The internal state will not be
```

Comment:
"...which the calculated hashes are written **to**."

Reply:
Hmm not so sure about that, there is already `into`.

Reply:
ahh, sorry, my bad

Reply:
I mistakenly assigned the `into` to the verb `passing` like "...are for passing a buffer into. which the calculated hashes are written to". So a comma would have helped me. But it's probably okay as it is. Don't know :smile:

Reply:
^^^ updated

Reply:
But there are still both `into` and `to` in your version.

Reply:
yea, it's just missing comma that got me confused. The RFC sentence makes sense now. My mistake was to put (in my mind) a comma behind `into`, and not already behind `buffer`.

Reply:
I can add that comma :)